feat(tools): Support string descriptions in Annotated parameters

Ratish1 opened the pull request:
## Description
<!-- Provide a detailed description of the changes in this PR --> This PR adds support for using typing.Annotated to define argument descriptions directly in the type hints of @tool decorated functions. This provides a cleaner and more maintainable way to document tool parameters.

To ensure a robust and maintainable implementation, this feature is currently focused on string-based descriptions. Support for `pydantic.Field` constraints within `Annotated` is deferred to a future enhancement, and a `NotImplementedError` is raised to guide developers.

### Key changes
- The tool decorator now correctly parses string metadata from `Annotated` type hints and uses it as the parameter's description, prioritizing it over docstrings.
- Uses `param.annotation` for type inspection to ensure consistent behavior across all supported Python versions, including Python 3.10.
- Adds a check to raise a `NotImplementedError` if `pydantic.Field` is used within an `Annotated` type.value handling, and correct priority.
- Includes unit tests to validate the new description parsing and the `NotImplementedError` case.

## Related Issues

<!-- Link to related issues using #issue-number format -->
#511 

## Documentation PR

<!-- Link to related associated PR in the agent-docs repo --> N/A

## Type of Change

<!-- Choose one of the following types of changes, delete the rest -->


New feature


## Testing

How have you tested the change?  Verify that the changes do not break functionality or introduce warnings in consuming repositories: agents-docs, agents-tools, agents-cli

- [x] I ran `hatch run prepare`
- [x] I ran formatters and linters
- [x] I ran the new unit tests:
- pytest -q tests/strands/tools/test_decorator.py → All 61 tests passed 

## Checklist
- [x] I have read the CONTRIBUTING document
- [x] I have added any necessary tests that prove my fix is effective or my feature works
- [ ] I have updated the documentation accordingly
- [ ] I have added an appropriate example to the documentation to outline the feature, or no new docs are needed
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Ratish1 commented:
The unit tests seems to be failing but they pass on my end, is it due to the logic written in decorator file __init__ method @dbschmigelski

dbschmigelski commented:
> The unit tests seems to be failing but they pass on my end, is it due to the logic written in decorator file **init** method @dbschmigelski

They are failing when I run locally. I think there are some small bugs related to truthiness.  I also am not seeing proper handling of the required case.

I think there is an issue when it comes to defaulting when we have Field within Annotated

with

```
        score: Annotated[int, Field(description="Score between 0-100", ge=0, le=100)] = 50,
```


Field does not have a default so it is incorrectly still in required,


If we did


```
        score: Annotated[int, Field(description="Score between 0-100", ge=0, le=100, default=20)] = 50,
```
it would pass though


So we have to address the failing tests and probably add a test with this inner-default

dbschmigelski commented:
I can check again regarding the branch protection.


 I think there are some interesting things going on for the test failures

```
FAILED tests/strands/tools/test_decorator.py::test_tool_decorator_annotated_pydantic_field_constraints - AssertionError: assert 'score' not in ['email', 'score']
FAILED tests/strands/tools/test_decorator.py::test_tool_decorator_annotated_no_description_fallback - KeyError: 'description'
```


For 
```
FAILED tests/strands/tools/test_decorator.py::test_tool_decorator_annotated_pydantic_field_constraints - AssertionError: assert 'score' not in ['email', 'score']
```

this is the default issue I mentioned, where when default is present INSIDE the Field it is respected. But, if we have an Annotation, with a Field, and the Annotation has the default then we unexpectedly are not respecting the default.



For 
```
FAILED tests/strands/tools/test_decorator.py::test_tool_decorator_annotated_no_description_fallback - KeyError: 'description'
```

It looks like when the Field does not have any description, even when we do `field.description = ...` it is not getting set properly which is strange

Ratish1 commented:
> I can check again regarding the branch protection.
> 
> I think there are some interesting things going on for the test failures
> 
> ```
> FAILED tests/strands/tools/test_decorator.py::test_tool_decorator_annotated_pydantic_field_constraints - AssertionError: assert 'score' not in ['email', 'score']
> FAILED tests/strands/tools/test_decorator.py::test_tool_decorator_annotated_no_description_fallback - KeyError: 'description'
> ```
> 
> For
> 
> ```
> FAILED tests/strands/tools/test_decorator.py::test_tool_decorator_annotated_pydantic_field_constraints - AssertionError: assert 'score' not in ['email', 'score']
> ```
> 
> this is the default issue I mentioned, where when default is present INSIDE the Field it is respected. But, if we have an Annotation, with a Field, and the Annotation has the default then we unexpectedly are not respecting the default.
> 
> For
> 
> ```
> FAILED tests/strands/tools/test_decorator.py::test_tool_decorator_annotated_no_description_fallback - KeyError: 'description'
> ```
> 
> It looks like when the Field does not have any description, even when we do `field.description = ...` it is not getting set properly which is strange

No I figured out the branch protection happens cause I opened the PR from my fork, I should be opening it from the main repo itself. So its my fault, if you want I can create a new PR immediately with the same changes.

Ratish1 commented:
Hi @dbschmigelski , Ik I said I couldnt figure it out, but I wanted to keep trying, so I created a new commit. I discovered that FieldInfo.default is PydanticUndefined when no default is set, which is different from an explicit default of None. Hopefully this works.

dbschmigelski commented on src/strands/tools/decorator.py:
```diff
@@ -97,23 +101,75 @@ def __init__(self, func: Callable[..., Any], context_param: str | None = None) -
+                    #
+                    # Deferring this complexity until there is clear demand and a robust
+                    # pattern for inspecting FieldInfo constraints is established.
+                    raise NotImplementedError(
```
👍 thanks so much this is great

dbschmigelski commented:
>it passes for me locally

Strange, I had thought hatch would prevent issues

Ratish1 commented:
> > it passes for me locally
> 
> Strange, I had thought hatch would prevent issues

I normally just do pre commit check all command. I tried hatch and it passes for me, dont know if this is a me problem. Anyway hopefully it works now.

Ratish1 commented:
It seems to still fail @dbschmigelski . I think get_type_hints() behaves differently for complex Annotated types in older Python versions so I removed it and switched to use param.annotation directly.

Ratish1 commented:
Wanted to follow up on this @dbschmigelski , I think it should work now, thanks.
